### ISSUE TITLE
Simplify hooks

### ISSUE THREAD

**Issue by stephenh** [stephenh/joist-ts#236]
Currently we have the following hooks:

* beforeDelete
* beforeFlush
* beforeCreate
* beforeUpdate
* afterValidation
* afterCommit

With a few thoughts:

* We copied the "before" / "after" terminology from Rails, but we really don't have "beforeFoo" and "afterFoo" for every "foo" stage, and it's not clear we should have that.
* We currently don't use `afterValidation`
* We have 3 `afterCommit`s that push data into external systems
* We've been tripped by `beforeFlush` running before `beforeCreate` b/c we want to apply defaults in `beforeCreate` that are seen by `beforeFlush`

Rough proposal, remove the `before`/`after` naming convention, and fix the ordering:

* `onDelete`
* `onCreate` (fired only on create, and before `onSave`)
* `onUpdate` (fired only on update, and before `onSave`)
* `onSave` (fired on either create or update, and after `onCreate` / `onUpdate`
* `onCommit` the new `afterCommit`

**Comment by stephenh** [comment]
I guess I'm not sure what to do about `afterValidation`--I'd like to remove it, but without it there is not a differentiation between:

* An `onSave` hook that helps put the entity into a valid state, and
* An `onSave` hook that gets to see the known-to-be-valid state (which `onCommit` gets to see too, but that's after we've committed to the db)

**Comment by mmahalwy** [comment]
Thoughts on: https://github.com/stephenh/joist-ts/issues/154#issuecomment-1014288755

**Comment by stephenh** [comment]
Hrm, yeah, that's interesting; I'd forgot / didn't realize that RoR runs `after_validation` before the rest ... which I guess implies the entities _must_ be valid to get to the `before_save`/`before_create`/etc. hooks?

I think for us, we pretty heavily use hooks to take maybe-not-yet-valid entities and turn them into valid entities.

So my original reason for filing #154 was not that I thought our hooks should not see invalid versions of entities, just that they defacto do see potentially-invalid versions of entities (which again I think is a feature), but the type system is not telling them that...

Another wrinkle here is hook ordering, i.e. maybe I have a `onSave` "hook 1" that helps make the entity valid, but then another `onSave` that wants to observe the "after hook 1 runs" version of the entity. Today we can't encode that level of inter-hook dependencies.

I'm not sure if we want to; maybe we do b/c it's very explicit. Or maybe we don't, and that's where the notion of "before" and "after" become useful, like "before" hooks see potentially-invalid entities and "after" hooks see "should-be-valid-now (or at least the 'before' hooks have run)" entities.

**Comment by zgavin** [comment]
> We copied the "before" / "after" terminology from Rails, but we really don't have "beforeFoo" and "afterFoo" for every "foo" stage, and it's not clear we should have that.

I don't think this is a problem.  Using 'on' as a prefix hides when these things actually take place.  It's obvious from the naming when `afterCommit` happens in the lifecycle of a flush, it is not obvious when `onCommit` happens.  Rails itself doesn't have perfect parity between before/after "foo", either.

> `onSave` (fired on either create or update, and after onCreate / onUpdate)

If we're moving away from the rails naming, why would we move to their naming for this?  We do not have a `save` operation.  We have `flush`, so imo it should stay `beforeFlush` or `onFlush`

> Hrm, yeah, that's interesting; I'd forgot / didn't realize that RoR runs after_validation before the rest ... which I guess implies the entities must be valid to get to the before_save/before_create/etc. hooks?

Yes, but rails handles validations differently.  They are explicitly field level instead of entity/record level and generally simpler, ie not intended to validate across relations.  Effectively, they could be considered 'sync' in our parlance. Rails also relies more heavily on DB defaults to ensure that values are present before validation.  There are `after_initialization`, `after_find` and `before_validation` hooks for the odd exceptions.

The original problem that hooks were solving, iirc, was populating defaults after `setPartial` calls. This is why they currently run before validations.

**Comment by stephenh** [comment]
> If we're moving away from the rails naming, why would we move to their naming for this?

Ah yeah, I didn't mean to pick save as their terminology, it just seemed like a good term for "either create or update --> save".

I was trying to brainstorm hook names just from scratch, and initially "beforeFlush" seemed kinda "what is that even?"; i.e. "we have create, update, delete...what/when does flush mean compared to those...".

But you're right, it does line up with `em.flush`.

> you can't use hooks ... to put invalid entities into the db since validations are the last thing to run

Do you mean that would be a desirable thing to do?

> imo i think what we want is to have a separate 'set the defaults' hook, then have a simple
> validation pass for missing required fields, then run existing hooks (which would be able to
> now rely on type guarantees), and finally run the full validation rules. 

That seems doable. Would "set the defaults" replace `beforeCreate`? Or probably not replace the hook, but most of our current `beforeCreate` usages would move to `applyDefaults`? Would `applyDefaults` be called only for create, or every time?

**Comment by stephenh** [comment]
I kinda wonder if `beforeFlush` would, to a new user, insinuate that it is called immediately before the SQL is put on the wire (which is what "flush" means to me), so if anything is closer to the current `afterValidation`, in that I think (again if I was just starting out) I would expect `beforeFlush` to see "perfect" / "as-is-going-into-the-db" versions of entities.

Not be something to "help out" creating the entity itself. Which, iirc you're right, was the original use of `beforeFlush` ... and actually maybe why we made the odd-to-us/me-later decision to have `beforeFlush` go before `beforeCreate`/`beforeUpdate`, b/c it let those hooks see a more complete version of the entity.

It kinda seems like we have two axis of interest:

1. Does the hook want to see pre-validation or post-validation version of the entity?
2. Does the hook want to run on only-create/only-update/both?

So like `beforeValidCreate` / `afterValidCreate`, `beforeValidUpdate` / `afterValidCreate`, `beforeValidSave` / `afterValidSave`. 

I'm not actually proposing those names, but just trying to think about the problem.

I suppose your proposal was make bullet 1 even more nuanced: before-required-validation (new hook), after-required-validation (existing hooks), and after-validation (current `afterValidation`).

**Comment by stephenh** [comment]
Other misc thoughts:

Per the brainstorming about "complete-ness of the entity" as a purpose/categorization of hooks, nearly all of our hooks are for creating side-effects within the domain model itself, they are all mutating the domain. So pedantically very few of them (only the `afterCommit` & `afterValidation`) see "complete" as in will-not-change-anymore.

Right now we don't know what fields a hook is mutating; it would be neat if we did, b/c we might then be able infer an order of hooks, i.e. hook A writes `book.title` and book B reads `book.title`, so run A before B, even if they are both `beforeCreate` hooks.

Granted, I might be a little worried about determining hook order every time on boot ... but we could also have like a codegen-ish process that was auto-ran on any `${entity}.ts` change, that boots up the domain model merely to introspect + calc the hooks/metadata, and then save the resulting info into an additional `src/entities/metadata2.ts` file that, for normal boots, would just be read as-is. I only mention this b/c I thought of it as a mitigation to a problem we don't even have yet, so this isn't anything we should anytime soon/hopefully ever.

It'd be nice if the `beforeCreate` / `afterCreate` could connotate something about "before valid" / "after valid", but I don't think as-is `afterCreate` really implies that. But maybe we could think of other `before`/`after` type prefixes that would insinuate that.

Per renaming `beforeFlush`, really all of the hooks run "during `em.flush`", so I'm more emboldened to rename it to be "on either create or update" aka `beforeSave`.

**Comment by stephenh** [comment]
I wonder if we could use field-level reactivity hints to replace the notion of "before create" / "before update". Like maybe there is just "before save", and something about the hint lets us know "only create" or "only update". Probably not, just thinking.

**Comment by stephenh** [comment]
If we add a `setDefaults` hook, ideally factories would call that first before they apply their "dumb" defaults.
